core/stats_handler.go

```diff
@@ -0,0 +1,203 @@
+package core
+
+import (
+	"context"
+	"time"
+
+	tpb "github.com/textileio/go-threads/api/pb"
+	hpb "github.com/textileio/textile/v2/api/hubd/pb"
+	mdb "github.com/textileio/textile/v2/mongodb"
+	"google.golang.org/grpc/stats"
+)
+
+type StatsHandler struct {
+	t *Textile
+}
+
+var _ stats.Handler = (*StatsHandler)(nil)
+
+var statsTimeout = time.Second * 10
+
+// HandleRPC accounts for customer usage across services.
+func (h *StatsHandler) HandleRPC(ctx context.Context, st stats.RPCStats) {
```

Comment:
`StatsHandler` allows you to track data in it's own context (Note: which is not available to other interceptors). However, it does _not_ allow you to return an error, i.e., stop the request. So, really we can just use this for tracking purposes... errors have to be returned from with the request handler or an interceptor. 

So then, what's the point of this? It's the only way we can collect wire message size info for tracking egress. The steam interceptors might sound appealing for this, but you'd have to pop the message from the stream before letting it continue to handlers.  

It's also used here to track reads and writes since we have access to the payloads for unary and stream handlers.

The downside of this approach is that we can't (currently) protect against a request the will _for the first time_ cause a free quota overage. That has to be handled _on the next request_. Keep in mind, this is only the case for egress, reads, and writes. Data storage is handled differently. I have an idea about how to enable this fail-fast behavior for egress, but will save that for next PR.

Reply:
What's the worst case here though, really? Request comes in, bumps quotas, and then the next one comes in... worst case isn't wild right?

Reply:
Not wild, but somebody could accrue a bunch of egress by pulling a huge file, which caps at the max bucket size of 4Gib. Not too bad but also not negligible.